### DISCUSSION TITLE
Idea: Runtime Optimization, Iterator Blocks and Compiler-generated State Machines

### DISCUSSION THREAD

**Discussion by AdamSobieski** [dotnet/csharplang#7629] [Language Ideas]
Hello. I would like to discuss some ideas towards enabling the optimization of iterator blocks and their resultant compiler-generated state machines.

```cs
public IEnumerable<int> Function1()
{
    yield return 1;
    yield return 2;
    yield return 3;
}
```

At first, towards enabling runtime optimizations, one might consider features which would surface in C# in a manner resembling:

```cs
public IQueryable<int> Function2()
{
    yield return 1;
    yield return 2;
    yield return 3;
}
```

Next, one might consider that these functions are transformed by compilers into state machines. So, techniques for runtime optimization might surface in C# in a manner resembling:

```cs
public IStateMachine<int> Function3()
{
    yield return 1;
    yield return 2;
    yield return 3;
}
```

When developers would utilize the state-machine-related interfaces as return types, this would signal that they desired to provide iteration while also providing inspectability for runtime optimizers.

As considered, state machines would provide read-only access to their states and transitions. `IStateMachine<T>` would extend `IEnumerable<T>` and `IStateMachine`. `IStateMachine` would have a property of type `IStateMachineProvider`.

Models of compiler-generated state machines might provide access to `System.Linq.Expressions.Expression` or Roslyn tree expression nodes representing the transformed program logic inside of the state machines. With these sorts of features, one can start to imagine exploring how to provide developers with runtime optimizations.

We could create extension methods on these interfaces, on `IStateMachine` and `IStateMachine<T>`, resembling LINQ. These extension methods would be used, in particular, when runtime optimizations were desired.

I am in particular interested in optimizations with respect to use cases like [YieldProlog](https://github.com/vslab/YieldProlog/) where iterator blocks and compiler-generated state machines are utilized extensively.

These topics are interesting to me and hopefully are of some interest to you. Has anyone else explored these topics? What do you think of these ideas?

**Comment by CyrusNajmabadi** [comment]
@AdamSobieski similar to related discussions you've brought up in the past, this seems to lack information on why this would be beneficial. 

Furthermore, it doesn't really explain what the language would do here.

**Comment by AdamSobieski** [comment]
@CyrusNajmabadi, hello. I'm not sure that I'd tend to agree with that. The last time that I opened a discussion here was in February of this year and six of us, yourself included, appeared to enjoy talking about [Metadata Attributes on Statements and Expressions](https://github.com/dotnet/csharplang/discussions/7008).

This discussion topic about _Runtime Optimization, Iterator Blocks and Compiler-generated State Machines_ could be even more interesting, more significant, and, potentially, more beneficial to the developer community.

I wanted to share these ideas as soon as they occurred and, as a result, I am still awaiting some emails about some of the technical particulars pertaining to the optimization of multiple interconnected, interrelated, interdependent compiler-generated state machines. The intended benefits would be speed- and performance-related.

The gist of the envisioned language features, what the language would do here, would be to allow developers to choose to use another return type for their iterator and generator functions. That is, `IEnumerable<T>` would be utilized for the existing functionalities and something new would be available from the language for the "`IQueryable<T>` analogue".

The envisioned return-type interface, `IStateMachine<T>` or `IQueryableStateMachine<T>`, would extend `IEnumerable<T>`, supporting iteration, but would also provide runtime inspectability for compiler-generated state machines from multiple modular methods, classes, or assemblies, allowing the inspection of states, transitions, and compiler-generated program logic therein.

**Reply by HaloFour** [reply]
Those are very vague statements with very vague claims of benefits.  I think to have a discussion you're going to have to establish the problem you're trying to solve and how this, whatever this is, serves to solve that problem.  I followed the link about `YieldProlog` to the github repo which did absolutely nothing to explain what it is or what it does, and I'm certainly not interested in doing a lot more research to facilitate a discussion I find dubious to begin with.

In general I view the state machine to be entirely an implementation detail, one that should not be introspected at runtime via a public API.  Doing so makes those implementation details public and would drastically impact the ability of C# (or any language) to refactor those implementations.

**Reply by CyrusNajmabadi** [reply]
> I'm not sure that I'd tend to agree with that.

I was referring to your prior posts on language integrated domain modeling, and language integrated rules. Both of which were quite vague, lacked good cases of the problems they were trying to solve, and did not seem to answer why we needed language changes for them, instead of just a library. 

In the future, I'd recommend that you stay by building a library first, and showing the value that comes from it. Then show how using that library is sometimes clunky or redundant for the client. Then show how a language feature improves that experience. 

This will help make things concrete instead of vague. And it will help motivate what sort of language feature is being requested, and why it would be a good thing to do.

**Reply by AdamSobieski** [reply]
@CyrusNajmabadi , @HaloFour , Have you two noticed that you have had to hand-select two old posts from over three years ago, from 2020, to attempt to rationalize or to explain what can be reasonably construed as a professional misconduct on @CyrusNajmabadi's part?

Have you two noticed that you just verbally slighted a third-party software project, `YieldProlog`, as not worth looking at, when all I said about it is that it and its users make extensive use of generator functions?

I would also tend to disagree that potential optimizations, perhaps correlates of method inlining, from the runtime processing of compiler-generated state machines are "dubious".

Have you two noticed that you have rushed to close and hide this issue, which would otherwise have been available to the broader community, and perhaps to your managers, as evidence of that professional misconduct?

> In general I view the state machine to be entirely an implementation detail, one that should not be introspected at runtime via a public API. Doing so makes those implementation details public and would drastically impact the ability of C# (or any language) to refactor those implementations.

Ok. Based on this, there won't be developer-available runtime optimizations for compiler-generated state machines in C# or, instead, such optimizations would be a JIT topic.

**Reply by CyrusNajmabadi** [reply]
@AdamSobieski this is the csharplang repo and discussions here are guided and addressed as appropriate. It's not an open forum for anyone to post anything they want without moderation. 

In the case of your posts, I am asking for more concrete information to offset the regular vague, handwavy explanations you tend to give. Posts in the past have had this issue, so I wanted to nip that problem in the bud here.  You are welcome to continue it in the future provided you can flesh out the ideas and requests a bit more. 

In other words, the onus is on you to make a constructive post that fits within the domain of this repo and which can actually be the basis of constructive discussion. Posts that don't meet that level will be moderated appropriately. 

Thanks!

**Reply by CyrusNajmabadi** [reply]
> I would also tend to disagree that potential optimizations, perhaps correlates of method inlining, from the runtime processing of compiler-generated state machines are "dubious".

There has been a lack of information on this. And specifically, you even mentioned you didn't have the data/documents to back this up yet. As such, the ideas are *admittedly* highly speculative on your part. 

When you have the information/examples/motivations ready, we can continue this. The same as with the prior discussions.  Thanks!

**Reply by CyrusNajmabadi** [reply]
Note: you might want to consider discord as a good place to discuss these ideas with the team and community ahead of time. Both groups can help outline for you the information that would be needed and guide you on how to make a successful discussion. 

https://discord.com/channels/143867839282020352/598678594750775301 and 
https://discord.com/channels/732297728826277939/732297994699014164

Are good places to start.

**Reply by HaloFour** [reply]
@AdamSobieski 

> Have you two noticed that you just verbally slighted a third-party software project, YieldProlog, as not worth looking at, when all I said about it is that it and its users make extensive use of generator functions?

Not in the slightest.  Could be an excellent project.  But it's not my responsibility to research what it is in order to figure out how it makes your vague statements make any sense.  That's your responsibility.

**Reply by AdamSobieski** [reply]
> Could be an excellent project.

@HaloFour, thank you for taking a minute to clarify. A ticklish matter is that [Jeff Thompson](http://www.thefirst.org/jefft/index.html), creator of the YieldProlog project, is one of the people that I invited, last week, to observe and to potentially participate in this discussion thread.

**Reply by CyrusNajmabadi** [reply]
@AdamSobieski You mentioned:

> I am in particular interested in optimizations

But you haven't even given a description of what issues actually exist today, and what sort of optimizations would be possible, viable, and how they would even work. 

You've vaguely jsut said that there will be interfaces (like IStateMachine) without describing what those interfaces actually provide, and vaguely indicated that this will make optimizations possible, without showing a single example of that in action. 

Basically, there's nothing really possible here.  It's like saying: "if c# exposed an IGC interface, we could make memory management faster".  There's just nothing to go off of there.  

For these discussions/ideas to have traction, we need more to go off of.  My recommendation is to do some exploration first, then come with some more *fleshed out* ideas that people can realistically consider and then sink their teeth into.

**Reply by AdamSobieski** [reply]
> how they would even work.

There are a number of tools in the toolbox with which to explore the matter of accelerating and optimizing modern logic programming in syntactically-convenient, iterator-heavy approaches (like those in the Python/C#/JS versions of YP... or even in some LINQ approaches).

These include, but are not limited to, graph-based selectors, pattern-matching, rules-based systems, other transformations of state machines into state machines, mappings of state machines into other constructs from other domains, and other techniques with which to work upon the contents of multiple interconnected, interrelated, interdependent compiler-generated state machines.

Modeled state machines would, importantly, provide representations of their transformed program logic utilizing one of the AST models of the .NET ecosystem.

On the one hand, with AST-based representations of iterator functions' logic available, one might consider that, with that information, one could readily inline and optimize them. On the other hand, complexities might arise from the (esoteric?) compiler-generated state-machine representations and their corresponding transformed program logic.

As this isn't a brainstorming forum, where one could broach and discuss one implementational approach, and then a second and third, I wanted to double-check with some others interested in automata and state machines before presenting, or proposing, a concrete approach here.

With respect to fleshing out these ideas, any others interested in these topics might want to choose to publish their solutions in a computer-science journal... and hopefully with an accompanying open-source project!

**Reply by CyrusNajmabadi** [reply]
@AdamSobieski when you have something more fleshed out, feel free to start a new discussion on that idea. Thanks.

**Comment by CyrusNajmabadi** [comment]
>  I am still awaiting some emails about some of the technical particulars pertaining to the optimization

As this is a suggestion to make language changes to support optimizations, but no information had been given in that, I'm going to close this out. 

If you can provide better specifics in the future, that would be appreciated. Thanks!
